clarity-ext-scripts/clarity_ext_scripts/covid/report_to_sminet.py

```diff
@@ -0,0 +1,132 @@
+from datetime import datetime
+import logging
+from sminet_client import SampleMaterial, SmiNetError
+from clarity_ext.extensions import GeneralExtension
+from clarity_ext_scripts.covid.utils import CtmrCovidSubstanceInfo
+from clarity_ext_scripts.covid.services.sminet_service import SmiNetService
+from clarity_ext_scripts.covid.services.knm_service import KNMSampleAccessor
+from clarity_ext_scripts.covid.services.knm_sminet_service import (
+    KNMSmiNetIntegrationService, IntegrationError, UnregisteredPatient)
+from clarity_ext_scripts.covid.partner_api_client import (
+    COVID_RESPONSE_POSITIVE, PartnerClientAPIException)
+
+
+UDF_TRUE = "Yes"
+logger = logging.getLogger(__name__)
+
+
+def should_report(substance):
+    """
+    A substance is ignored from SmiNet reporting if it's a control, has been run before or doesn't
+    have a positive result. It is also ignored if the research engineer has set the UDF
+    "SmiNet status" to "ignore"
+
+    Samples that should be reported must be in the SmiNet state `retry` or `error` or None
+    """
+    if substance.is_control:
+        return False
+
+    if substance.submitted_sample.udf_rtpcr_covid19_result_latest != COVID_RESPONSE_POSITIVE:
+        return False
+
+    if substance.sminet_status in [SmiNetService.STATUS_IGNORE, SmiNetService.STATUS_SUCCESS]:
+        return False
+
+    if substance.sminet_status not in [
+            None,
+            SmiNetService.STATUS_RETRY,
+            SmiNetService.STATUS_ERROR]:
+        raise AssertionError(
+            "Unexpected SmiNet status: {}".format(substance.sminet_status))
+
+    return True
+
+
+class Extension(GeneralExtension):
+    """
+    Reports results to SmiNet.
+
+    For all analytes in the step:
+    * If sminet_status is not set:
+        * Checks if it should be imported.
+            * If it shouldn't, updates it to "ignore"
+            * If it should, tries to send the data
+    * If sminet_status is "error", retries it
+    * If sminet_status is "ignore", ignores it
+
+
+    Test data required:
+    * Positive  => success
+    * Negative  => ignore
+    * Failed    => ignore
+    * Anonymous => ignore
+    """
+
+    def report(self, substance):
+        """
+        Reports this substance to SmiNet and updates the status in the LIMS.
+        """
+
+        org_referral_code = substance.submitted_sample.name.split("_")[0]
+        date_arrival = substance.submitted_sample.api_resource.date_received
+        date_arrival = datetime.strptime(date_arrival, "%Y-%m-%d")
+
+        sample = KNMSampleAccessor(substance.submitted_sample.udf_knm_org_uri,
+                                   org_referral_code,
+                                   date_arrival,
+                                   SampleMaterial.SVALG)
+
+        integration = KNMSmiNetIntegrationService(self.config)
+        lab_result = SmiNetService.create_scov2_positive_lab_result()
+
+        error_msg = ""
+
+        # TODO: Anonymous should be ignored
```

Comment:
Just checking that this has not been forgotten.

Reply:
No it hasn't, forgot to remove the TODO (still in my staged changes)!